### ISSUE TITLE
False-negative possibly unbound variable with nested functions

### ISSUE THREAD

**Issue by cdce8p** [microsoft/pyright#4100]
**To Reproduce**
```py
from typing import Callable

def func(var: int) -> Callable[[], None]:
    if var == 0:
        state = 0
    elif var == 2:
        state = 1

    # print(state)

    def nested() -> None:
        print(state)  # state should be marked as possibly unbound

    return nested
```
`state` inside `nested` is only marked as `possibly unbound` if another expression inside `func` tries to access `state` first.
With the first print statement, pyright correctly detects both.
```
  test.py:9:11 - error: "state" is possibly unbound (reportUnboundVariable)
  test.py:12:15 - error: "state" is possibly unbound (reportUnboundVariable)
```

**Expected behavior**
`state` inside `nested` should be marked regardless of if the first print is present or not.

**VS Code extension or command-line**
1.1.277

**Comment by erictraut** [comment]
Unbound variable analysis is performed only locally within an execution context. Captured variables have lifetimes that generally cannot be analyzed statically. So the current behavior is expected.
